dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp

```diff
@@ -152,30 +146,66 @@ SolverKINAlgRestoration::initVarAndEqTypes() {
     }
   }
 
+  assert(numF_ == indexY_.size());
+
   if (ignoreF_.size() != ignoreY_.size() || indexF_.size() != indexY_.size()) {
 #ifdef _DEBUG_
     for (int i = 0; i < model_->sizeF(); ++i) {
       std::string fEquation("");
       std::string subModelName;
       int localFIndex = -1;
       model_->getFInfos(i, subModelName, localFIndex, fEquation);
-      Trace::debug() << DYNLog(SolverEquationsType, i, ((fType_[i] > 0)? "differential":"algebraic"), fEquation) << Trace::endline;
+      Trace::debug() << DYNLog(SolverEquationsType, i, ((modelFType[i] > 0)? "differential":"algebraic"), fEquation) << Trace::endline;
     }
     for (int i = 0; i < model_->sizeY(); ++i) {
-      Trace::debug() << DYNLog(SolverVariablesType, model_->getVariableName(i), i, ((vId_[i] > 0)? "differential":"algebraic")) << Trace::endline;
+      Trace::debug() << DYNLog(SolverVariablesType, model_->getVariableName(i), i, ((modelYType[i] > 0)? "differential":"algebraic")) << Trace::endline;
     }
 #endif
     throw DYNError(Error::SOLVER_ALGO, SolverUnbalanced);
   }
+
+  if (numFPrevious != numF_)
+    return true;
+  else
+    return false;
 }
+
 void
-SolverKINAlgRestoration::modifySettings(double fnormtol, double initialaddtol, double scsteptol, double mxnewtstep,
-                  int msbset, int mxiter, int printfl) {
-  if (nbF_ == 0)
+SolverKINAlgRestoration::setupNewAlgebraicRestoration(double fnormtol, double initialaddtol, double scsteptol, double mxnewtstep,
+                                  int msbset, int mxiter, int printfl) {
+  bool initKinsol = initVarAndEqTypes();
+  if (numF_ == 0)
     return;
+  if (initKinsol) {
+    // warning: model_->sizeF() != nbF_
+    // model_->sizeF() is fixed during the whole simulation
+    // nbF_ could vary
+    vectorF_.resize(model_->sizeF());
+
+    vectorYOrYpSolution_.assign(numF_, 0.);
+    cleanAlgebraicVectors();
+    sundialsVectorY_ = N_VMake_Serial(numF_, &(vectorYOrYpSolution_[0]));
+
+    if (sundialsVectorY_ == NULL)
+      throw DYNError(Error::SUNDIALS_ERROR, SolverCreateYY);
+
+    clean();
+    switch (mode_) {
+      case KIN_NORMAL:
+        initCommon("KLU", fnormtol, initialaddtol, scsteptol, mxnewtstep, msbset, mxiter, printfl, evalF_KIN, evalJ_KIN, sundialsVectorY_);
+        break;
+      case KIN_YPRIM:
+        initCommon("KLU", fnormtol, initialaddtol, scsteptol, mxnewtstep, msbset, mxiter, printfl, evalF_KIN, evalJPrim_KIN, sundialsVectorY_);
+        break;
+    }
+  } else {
+    updateKINSOLSettings(fnormtol, initialaddtol, scsteptol, mxnewtstep, msbset, mxiter, printfl);
```

Comment:
I think it's a detail but it is to be sure that I correctly understand the new behavior.
In the implementation you propose here, if I have two algebraic restoration without any changes in the types of f, you update the settings while you don't have too, no?

Reply:
Yes that is the new behavior because no there is no memory of such info anymore and I think it would complicate things for no reason